### DISCUSSION TITLE
Should we mask potentially sensitive information in `sys.jobs` and `sys.jobs_log`?

### DISCUSSION THREAD

**Discussion by mfussenegger** [crate/crate#10300] [General] [answered]
There are a couple of statements that can potentially contain confidential information. To mind come `CREATE USER` - which can include the users password, and `COPY FROM` / `COPY TO` - which could include the access key and secret for S3.

By default the statements show up in `sys.jobs`, and also appear in `sys.jobs_log` after execution.
Users only see their own statements when they query the tables. User A doesn't see statements that user B executed. The exception to this is the `crate` super-user, which can see all statements.

This could be seen as a security issue. From [Logging cheat sheet](https://cheatsheetseries.owasp.org/cheatsheets/Logging_Cheat_Sheet.html#data-to-exclude):

> "Access tokens" are among the data that "should not usually be recorded directly in the logs, but instead should be removed, masked, sanitized, hashed or encrypted".

There are currently several options to lock this down:

1. Disable jobs logs completely via `SET GLOBAL "stats.enabled" = false`
2. Take care and don't inline sensitive values, but use parameter substitution instead. (Though we are considering to also expose args in the jobs log, so this would then no longer be an option)
3. Set a custom `stats.jobs_log_filter` to exclude statements that could contain sensitive information. But this would only affect `sys.jobs_log`, not `sys.jobs`.

Both point one and three rely on trusting the system administrator, because only the `crate` user can change these options. 

We could attempt to mask sensitive values in queries. This would work well for queries like `CREATE USER` or `COPY`, but only if the statements can be parsed. If there would be a typo, the statement would still show up in `sys.jobs_log`.
If other statements like `INSERT INTO` contain data that is considered confidential, that approach would also not work, unless we mask by default *any* value. Again. This wouldn't work for statements that cannot be parsed.

So the questions is: Should we attempt to mask values, or should we just point out the current behavior in the documentation?

**Comment by seut** [comment, marked as answer]
As you pointed out, masking would only work quite limited, it would not work on parameter values nor on non-parsable statements (typos).
Given that, I'd rather improve out documentation to point out the current default behavior and also options to prevent it (like you listed) than implementing somehow "half-baked" solution which also increases complexity.
Maybe we could also add features like a `sys.jobs_filter` to help administrators to improve security on their own?

**Comment by MarkusH** [comment]
I've also looked into PostgreSQL's query logging. And they, even when writing to a file on disk, don't mask anything in a `CREATE USER` statement.

Masking will never be 100% safe and secure, but it can reduce the exposure. If it's not too complex, adding some safeguards still seems like a good idea to me. And yeah, that would only work with valid queries.

On the up-side, my understanding is that `sys.jobs_log` is also only in-memory and once a node is restarted, its jobs log is gone.

**Reply by mfussenegger** [reply]
> On the up-side, my understanding is that sys.jobs_log is also only in-memory and once a node is restarted, its jobs log is gone.

Yes that is true, unless a `stats.jobs_log_persistent_filter` is set.


Will keep this open for a bit longer to see if there is other inputs. Otherwise will opt for *not* masking but documenting the potential exposure.
